os/binfmt/libelf/libelf_bind.c

```diff
@@ -441,42 +425,19 @@ int elf_bind(FAR struct elf_loadinfo_s *loadinfo, FAR const struct symtab_s *exp
 		}
 	}
 
-#if defined(CONFIG_ARCH_ADDRENV)
+#if defined(CONFIG_ARCH_HAVE_COHERENT_DCACHE)
 	/* Ensure that the I and D caches are coherent before starting the newly
 	 * loaded module by cleaning the D cache (i.e., flushing the D cache
 	 * contents to memory and invalidating the I cache).
 	 */
 
-#if 0							/* REVISIT... has some problems */
-	(void)up_addrenv_coherent(&loadinfo->addrenv);
-#else
-	up_coherent_dcache(loadinfo->textalloc, loadinfo->textsize);
-	up_coherent_dcache(loadinfo->dataalloc, loadinfo->datasize);
-#endif
-
-	/* Restore the original address environment */
-
-	status = elf_addrenv_restore(loadinfo);
-	if (status < 0) {
-		berr("ERROR: elf_addrenv_restore() failed: %d\n", status);
-		if (ret == OK) {
-			ret = status;
-		}
-	}
-#elif defined(CONFIG_ARCH_HAVE_COHERENT_DCACHE)
-	/* Ensure that the I and D caches are coherent before starting the newly
-	 * loaded module by cleaning the D cache (i.e., flushing the D cache
-	 * contents to memory and invalidating the I cache).
-	 */
-
-	up_coherent_dcache(loadinfo->textalloc, loadinfo->textsize);
-	up_coherent_dcache(loadinfo->dataalloc, loadinfo->datasize);
+	up_coherent_dcache(loadinfo->binp->sections[BIN_TEXT], loadinfo->binp->sizes[BIN_TEXT]);
+	up_coherent_dcache(loadinfo->binp->sections[BIN_DATA], loadinfo->binp->sizes[BIN_DATA]);
 
 #endif
 
-#if defined(CONFIG_ARCH_ADDRENV) || defined(CONFIG_SUPPORT_COMMON_BINARY)
+#ifdef CONFIG_SUPPORT_COMMON_BINARY
 ret_err:
```

Comment:
Originally, `strtab` and `symtab` will be freed regardless of `ret_err`.
But with your changes, `strtab` and `symtab` will be freed only when `CONFIG_SUPPORT_COMMON_BINARY` is enabled.
Could you check this?